### PR TITLE
[chore] refactore carbonexporter to handle errors, use generic interfaces

### PR DESCRIPTION
**Pull request by bogdandrutu**

### CODE REVIEW

File: exporter/carbonexporter/exporter.go [hunk 1]
```diff
@@ -109,6 +99,18 @@ func (cp *connPool) Write(bytes []byte) (int, error) {
 		}
 	}
 
+	// The deferred function below is what puts back connections on the pool if no error.
+	defer func() {
```
Comment (bogdandrutu) [inline exporter/carbonexporter/exporter.go:103]: Simplify logic, conn cannot be nil

File: exporter/carbonexporter/exporter.go [hunk 2]
```diff
@@ -137,20 +139,22 @@ func (cp *connPool) Write(bytes []byte) (int, error) {
...
 
-func (cp *connPool) Close() {
+func (cp *connPool) Close() error {
 	cp.mtx.Lock()
 	defer cp.mtx.Unlock()
 
+	var errs error
 	for _, conn := range cp.conns {
-		conn.Close()
+		errs = multierr.Append(errs, conn.Close())
```
Comment (atoulme) [inline exporter/carbonexporter/exporter.go:148]: can we use the `errors.Join` method here?

Comment (bogdandrutu) [reply exporter/carbonexporter/exporter.go:148]: I can if I really want, but is not as performant as append for this use-case. It replaces nicely `multierr.Combine` but not the append.

Comment (TylerHelmuth) [reply exporter/carbonexporter/exporter.go:148]: We have https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25121 to track switching to `errors.Join`.

Comment (bogdandrutu) [reply exporter/carbonexporter/exporter.go:148]: So will do in a separate PR then when we solve that.

**Review by dmitryax — APPROVED** [review]

**Pull request merged**